[JSC] Always initialize result array in wasm return frame

syg opened the pull request:
#### 9aa33190dfc1d30ec09060b3d12ec2b5337e86f4
<pre>
[JSC] Always initialize result array in wasm return frame

<a href="https://bugs.webkit.org/show_bug.cgi?id=297812">https://bugs.webkit.org/show_bug.cgi?id=297812</a>
<a href="https://rdar.apple.com/158847247">rdar://158847247</a>

Reviewed by Keith Miller.

Allocating uninitialized JSArrays is faster but unsafe. For use in Wasm
multi-value returns, initializing the return array can call into user code and
throws, making use of uninitialized JSArrays error-prone as all exception
return paths have to take care to ensure that array gets fully initialized.

This PR changes the Wasm multi-value return use case to use initialized arrays.
This change is performance neutral on JetStream and Speedometer.

Canonical link: <a href="https://commits.webkit.org/299210@main">https://commits.webkit.org/299210@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7681e2b803b295e877089c694df6aa7321374c6c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118152 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37830 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124307 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70192 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ab58f3ca-170d-4686-9251-c4284555e9b2) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46412 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89661 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59290 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/449f173d-8926-4ea1-a3b9-88a60e2322b1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30683 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105936 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70154 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9f4d1e2d-0fe2-465a-aa10-b73f5cf7b08e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29751 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24052 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67975 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110263 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100109 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24229 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127384 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116662 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45055 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33958 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98339 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45416 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102157 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98125 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24969 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43520 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21511 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41520 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44927 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50601 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145360 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44387 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37394 "Found 1 new JSC binary failure: testapi, Found 19702 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Basics/flags.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47732 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46076 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/ObjectInitializationScope.cpp:
```diff
@@ -65,13 +65,23 @@ void ObjectInitializationScope::notifyInitialized(JSObject* object)
+
 void ObjectInitializationScope::verifyPropertiesAreInitialized(JSObject* object)
 {
+    if (!object->butterfly())
```
We should probably ASSERT this object doesn't have inline properties or they're initialized. Can we add (somewhere): 

```
// ObjectInializationScope doesn't support objects with inline storage right now. If we wanted to support this we should have a way to visitChildren on the object with a debug `AbstractSlotVisitor` that checks the outgoing references are valid.
ASSERT(object->cellContainter().cellSizeForDebugging() == sizeof(JSObject));
```
In theory you could do some rounding to make sure it matches a size class for `JSObject` but that's almost certainly going to have a precise fit size class forever. You'll also need to add a 

```
CellContainer::cellSizeForDebugging() const { return isMarkedBlock() ? markedBlock().cellSize() : preciseAllocation.cellSize(); }
```

syg replied:
Comment now obsolete as going in the direction of using initialized arrays.

kmiller68 reviewed (approved):
r=me

syg committed:
[JSC] Always initialize result array in wasm return frame

https://bugs.webkit.org/show_bug.cgi?id=297812
rdar://158847247

Reviewed by Keith Miller.

Allocating uninitialized JSArrays is faster but unsafe. For use in Wasm
multi-value returns, initializing the return array can call into user code and
throws, making use of uninitialized JSArrays error-prone as all exception
return paths have to take care to ensure that array gets fully initialized.

This PR changes the Wasm multi-value return use case to use initialized arrays.
This change is performance neutral on JetStream and Speedometer.

Canonical link: https://commits.webkit.org/299210@main
